mathematica/mathematica.hpp

```diff
@@ -25,90 +28,146 @@ using geometry::R3Element;
 using geometry::Vector;
 using numerics::FixedVector;
 using physics::DegreesOfFreedom;
+using quantities::Amount;
+using quantities::Angle;
+using quantities::Current;
+using quantities::Exponentiation;
+using quantities::is_tuple_v;
+using quantities::Length;
+using quantities::LuminousIntensity;
+using quantities::Mass;
+using quantities::Pow;
 using quantities::Quantity;
 using quantities::Quotient;
+using quantities::Temperature;
+using quantities::Time;
+
+// A helper class for type erasure of quantities.  It may be used with the
+// functions in this file to remove the dimensions of quantities (we know that
+// Mathematica is sluggish when processing quantities).  Usage:
+//
+//   ToMathematica(... , ExpressIn(Metre, Second, Degree));
+//
+// The construction parameters must be for the base SI units.  They may be in
+// any order.  If not enough parameters are specified for complete type erasure
+// of the argument, compilation fails.  An object with no template parameters
+// may be used to indicate that type erasure should not happen.
+template<typename... Qs>
+class ExpressIn {
+ public:
+  static constexpr bool is_default = sizeof...(Qs) == 0;
+
+  // Check that only base SI units are specified.  If a unit is specified
+  // multiple times, the calls to std::get in operator() won't compile.
+  static_assert(
+      is_default ||
+      ((std::is_same_v<Qs, Length> || std::is_same_v<Qs, Mass> ||
+        std::is_same_v<Qs, Time> || std::is_same_v<Qs, Current> ||
+        std::is_same_v<Qs, Temperature> || std::is_same_v<Qs, Amount> ||
+        std::is_same_v<Qs, LuminousIntensity> || std::is_same_v<Qs, Angle>) ||
+       ...));
```

Comment:
Good point, I got the logic wrong (once more).

The fold expression that you write doesn't work with an empty pack, it would need a `... && true` at the end.

More importantly, `is_default` is important for compilation performance: the default is used all over the place, and compilation is really slower if this condition cannot be short-circuited.

Reply:
Done.